cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractProgramController.java

```diff
@@ -416,94 +416,104 @@ private ListenerCaller(Listener listener, Executor executor) {
       this.executor = executor;
     }
 
+    private void execute(Runnable runnable, String methodName) {
+      try {
+        executor.execute(runnable);
+      } catch (Throwable t) {
```

Comment:
This only works if the executor is a same thread executor. I think it is better do it in the MultiListenerCaller methods